src/AggregateFunctions/QuantileTDigest.h

```diff
@@ -128,71 +170,89 @@ class TDigest
       */
     void compress()
     {
-        if (unmerged > 0)
+        if (unmerged > 0 || centroids.size() > params.max_centroids)
         {
+            // unmerged > 0 implies centroids.size() > 0, hence *l is valid below
             RadixSort<RadixSortTraits>::executeLSD(centroids.data(), centroids.size());
 
-            if (centroids.size() > 3)
-            {
-                /// A pair of consecutive bars of the histogram.
-                auto l = centroids.begin();
-                auto r = std::next(l);
+            /// A pair of consecutive bars of the histogram.
+            auto l = centroids.begin();
+            auto r = std::next(l);
 
-                Count sum = 0;
-                while (r != centroids.end())
+            const BetterFloat count_epsilon_4 = count * params.epsilon * 4; // Compiler is unable to do this optimization
+            BetterFloat sum = 0;
+            BetterFloat l_mean = l->mean; // We have high-precision temporaries for numeric stability
+            BetterFloat l_count = l->count;
+            while (r != centroids.end())
+            {
+                if (l->mean == r->mean) // Perfect aggregation (fast). We compare l->mean, not l_mean, to avoid identical elements after compress
```

Comment:
Won't we have problems with floating point numbers comparison? Maybe better to compare their difference with some comparison epsilon?